numpy/testing/_private/utils.py

```diff
@@ -719,32 +734,23 @@ def chk_same_position(x_id, y_id, hasval='nan'):
             raise AssertionError(msg)
 
         if isnumber(x) and isnumber(y):
-            has_nan = has_inf = False
+            xy_isnan = xy_isinf = False
             if equal_nan:
-                x_isnan, y_isnan = isnan(x), isnan(y)
-                # Validate that NaNs are in the same place
-                has_nan = any(x_isnan) or any(y_isnan)
-                if has_nan:
-                    chk_same_position(x_isnan, y_isnan, hasval='nan')
+                xy_isnan = func_same_pos(isnan, x, y, hasval='nan')
 
             if equal_inf:
-                x_isinf, y_isinf = isinf(x), isinf(y)
-                # Validate that infinite values are in the same place
-                has_inf = any(x_isinf) or any(y_isinf)
-                if has_inf:
-                    # Check +inf and -inf separately, since they are different
-                    chk_same_position(x == +inf, y == +inf, hasval='+inf')
-                    chk_same_position(x == -inf, y == -inf, hasval='-inf')
-
-            if has_nan and has_inf:
-                x = x[~(x_isnan | x_isinf)]
-                y = y[~(y_isnan | y_isinf)]
-            elif has_nan:
-                x = x[~x_isnan]
-                y = y[~y_isnan]
-            elif has_inf:
-                x = x[~x_isinf]
-                y = y[~y_isinf]
+                xy_isinf = func_same_pos(lambda xy: xy == +inf, x, y,
+                                         hasval='+inf')
+                xy_isinf |= func_same_pos(lambda xy: xy == -inf, x, y,
+                                          hasval='-inf')
+
+            if xy_isnan is not False:
+                if xy_isinf is not False:
+                    x, y = remove_flagged(x, y, xy_isinf | xy_isnan)
+                else:
+                    x, y = remove_flagged(x, y, xy_isnan)
+            elif xy_isinf is not False:
+                x, y = remove_flagged(x, y, xy_isinf)
```

Comment:
Seems like micro-optimization here - can't we just write in place of these 7 lines:
```
to_remove = xy_isinf | xy_isnan
x = x[~to_remove]
y = y[~to_remove]
```

Reply:
Doesn't work for the same reason.

Reply:
The problem was that `xy_isinf` could just be `False`, in which case the flags would not broadcast correctly. But of course, this is easily checked in `remove_flagged`, so I did that now.